pkg/nodeauth/jwt/node_jwt_validator.go

```diff
@@ -0,0 +1,186 @@
+		return nil, fmt.Errorf("invalid claims type")
+	}
+	return claims, nil
+}
+
+// decodeP2PId converts string-encoded p2pId in the JWT claim to PeerID(ed25519 public key)
+func (v *NodeJWTAuthenticator) decodeP2PId(p2pIdString string) (p2ptypes.PeerID, error) {
+	var peerID p2ptypes.PeerID
+	err := peerID.UnmarshalText([]byte(p2pIdString))
+	if err != nil {
+		return p2ptypes.PeerID{}, fmt.Errorf("failed to unmarshal PeerID: %w", err)
+	}
+	return peerID, nil
+}
+
+// verifyJWTSignature prove the JWT signature is signed by the node's private key.
+// DEV NOTE: The designed JWT method is asymmetric Ed25519, @nodeauth/node_jwt_signer.go.
+func (v *NodeJWTAuthenticator) verifyJWTSignature(tokenString string, p2pId p2ptypes.PeerID) error {
+
+	token, err := v.parser.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		// Ensure the signing method is Ed25519
+		if _, ok := token.Method.(*jwt.SigningMethodEd25519); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+
+		// Convert PeerID to ed25519.PublicKey for signature verification
+		// PeerID is a [32]byte array that represents an ed25519 public key
+		publicKey := ed25519.PublicKey(p2pId[:])
+
+		// DEV NOTE: Internally, the JWT parser's callback use the returned value here to verify the signature.
+		return publicKey, nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("signature verification failed: %w", err)
+	}
+
+	if !token.Valid {
+		return fmt.Errorf("token is invalid after signature verification")
+	}
+
+	return nil
+}
+
+// verifyRequestDigest ensures the request hasn't been tampered with by verifying the digest in the JWT claim.
+func (v *NodeJWTAuthenticator) verifyRequestDigest(claims *NodeJWTClaims, originalRequest any) error {
+	// Calculate expected digest using the same method as the client
+	expectedDigest := utils.CalculateRequestDigest(originalRequest)
+
+	if claims.Digest != expectedDigest {
+		return fmt.Errorf("digest mismatch: got %s, want %s", claims.Digest, expectedDigest)
+	}
+
+	return nil
+}
+
+// verifyStandardClaims validates standard JWT fields.
+// DEV NOTE: This does not perform signature (public vs private key) verification.
+func (v *NodeJWTAuthenticator) verifyStandardClaims(claims *NodeJWTClaims) error {
+	now := time.Now()
+
+	// Check basic type-safety of struct fields
+	if err := v.validator.Struct(claims); err != nil {
+		return fmt.Errorf("claims validation failed: %w", err)
+	}
+
+	// Check expiration
+	if claims.ExpiresAt != nil && claims.ExpiresAt.Before(now) {
+		return fmt.Errorf("token expired at %v", claims.ExpiresAt.Time)
+	}
+
+	// Check not before (if set)
+	if claims.NotBefore != nil && claims.NotBefore.After(now) {
+		return fmt.Errorf("token not valid until %v", claims.NotBefore.Time)
+	}
+
+	// Check issued at
+	if claims.IssuedAt != nil && claims.IssuedAt.After(now.Add(workflowJWTExpiration)) {
+		return fmt.Errorf("token issued too far in future: %v", claims.IssuedAt.Time)
+	}
```

Comment:
This can be done by the JWT library - WithIssuedAt and similar [ParserOptions](https://pkg.go.dev/github.com/golang-jwt/jwt/v5#ParserOption)